File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        left = right = head

        for _ in range(n):
            right = right.next

        # If right is None, it means n is equal to the length of the list.
        # The node to remove is the head itself.
        if not right:
            return head.next

        guy_n_steps_behind = left
        guy_n_steps_ahead = right

        while guy_n_steps_ahead and guy_n_steps_ahead.next:
            guy_n_steps_behind = guy_n_steps_behind.next
            guy_n_steps_ahead = guy_n_steps_ahead.next

        # at this point, the two pointers that are n steps apart, have reached the end (the rightmost one has reached the end)
        # and the ListNode we wanna delete (basically skip) is the one just one step after the leftmost pointer. (guy_n_steps_behind.next)

        to_be_skipped = guy_n_steps_behind.next
        new_next = to_be_skipped.next

        guy_n_steps_behind.next = new_next

        return head





        

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Create a dummy node that points to head.
        // This will help us avoid special cases when deleting the head itself.
        ListNode dummy(0);
        dummy.next = head;

        // Both pointers start at dummy
        ListNode* left = &dummy;
        ListNode* right = &dummy;

        // Then we move the right pointer n+1 steps ahead
        // To create a gap of exactly n nodes between left and right
        for (int i = 0; i <= n; i++) {
            right = right->next;
        }

        // Then we move pointers together now until right reaches the end
        // Since the gaps here will now stay constnat, left will end up 
        // immediately BEFORE the node we want to remove (technically skip))
        while (right) {
            left = left->next;
            right = right->next;
        }
        
        // this would have skipped over the target node

        left->next = left->next->next; // skip

        return dummy.next;
    }
};